optaplanner-website-root/content/learn/drl-to-constraint-streams-migration.adoc

```diff
@@ -0,0 +1,447 @@
+
+#### Solver configuration
+
+Users of Quarkus most likely do not need to worry about solver configuration.
+Removing score DRL and introducing a `ConstraintProvider` implementation ought to be enough.
+
+In solver configuration XML however, score DRL is configured by pointing the solver to the DRL file:
+
+[source, xml]
+----
+<solver>
+
+  ...
+
+  <scoreDirectorFactory>
+    <scoreDrl>org/optaplanner/examples/vehiclerouting/optional/score/vehicleRoutingConstraints.drl</scoreDrl>
+  </scoreDirectorFactory>
+
+  ...
+
+</solver>
+----
+
+Constraint Streams are selected by pointing the solver to an implementation of the `ConstraintProvider` interface:
+
+[source, xml]
+----
+<solver>
+
+  ...
+
+  <scoreDirectorFactory>
+    <constraintProviderClass>org.optaplanner.examples.vehiclerouting.score.VehicleRoutingConstraintProvider</constraintProviderClass>
+  </scoreDirectorFactory>
+
+  ...
+
+</solver>
+----
+
+### Migrating trivial constraints
+
+Many constraint follow a simple pattern of picking an entity and immediately penalizing it.
+One such case can be found in the Vehicle Routing example:
+
+[source, drl]
+----
+rule "distanceToPreviousStandstill"
+    when
+        Customer(previousStandstill != null, $distanceFromPreviousStandstill : distanceFromPreviousStandstill)
+    then
+        scoreHolder.addSoftConstraintMatch(kcontext, - $distanceFromPreviousStandstill);
+end
+----
+
+Here, each initialized `Customer` instance incurs a soft penalty equivalent to the value of its `distanceFromPreviousStandstill` field. Here's how the same is achieved in CS:
+
+[source, java]
+----
+Constraint distanceToPreviousStandstill(ConstraintFactory factory) {
+    return factory.forEach(Customer.class)
+        .penalizeLong("distanceToPreviousStandstill",
+            HardSoftLongScore.ONE_SOFT,
+            customer -> customer.getDistanceFromPreviousStandstill());
+}
+----
+
+Note that:
+
+* `forEach(Customer.class)` serves the same purpose as `Customer(...)` in DRL.
+* There is no need to check if planning variable is initialized (`previousStandstill != null`), as `forEach(...)` does it automatically.
+If this behavior is not what you want, use `forEachIncludingNullVars(...)` instead.
+* Right-hand side of the rule (the part after `then`) is replaced by a call to `penalizeLong(...)`.
+The size of the penalty is now determined by constraint weight (`HardSoftLongScore.ONE_SOFT`)
+and match weight (the call to a getter on `Customer`).
+
+#### Applying rewards instead of penalties
+
+In the example above, score DRL applies a penalty by adding a negative constraint match like so:
+
+`scoreHolder.addSoftConstraintMatch(kcontext, - $distanceFromPreviousStandstill)`.
+
+CS makes this more explicit by using a keyword `reward` instead of `penalize`, while keeping the match weight positive:
+
+`rewardLong(..., ..., customer -> customer.getDistanceFromPreviousStandstill())`.
+
+#### Applying different penalty types
```

Comment:
Before this part, discuss that impact = constraint weight multiplied by match weight.
discuss that the constraint weight determines both the constraint level and a base weight, equal for all matches of this constraint, but that a match weigh is match specific.